pkg/agent/manager/cache/cache.go

```diff
@@ -107,37 +109,52 @@ func (c *cacheImpl) SetEntry(entry *Entry) {
 	c.cache[entry.RegistrationEntry.EntryId] = entry
 	c.m.Unlock()
 
-	subs := c.Subscribers.Get(entry.RegistrationEntry.Selectors)
+	subs := c.subscribers.get(entry.RegistrationEntry.Selectors)
 	c.notifySubscribers(subs)
 }
 
-func (c *cacheImpl) notifySubscribers(subs []*Subscriber) {
-	go func() {
-		entries := c.Entries()
-		bundle := c.Bundle()
-		for _, sub := range subs {
-			subEntries := subscriberEntries(sub, entries)
-			select {
-			case <-sub.done:
-				c.Subscribers.remove(sub)
-			case sub.C <- &WorkloadUpdate{Entries: subEntries, Bundle: bundle}:
-			}
+func (c *cacheImpl) notifySubscribers(subs []*subscriber) {
+	if subs == nil {
+		return
+	}
+
+	c.notifyMutex.Lock()
+	defer c.notifyMutex.Unlock()
+
+	entries := c.Entries()
+	bundle := c.Bundle()
+	for _, sub := range subs {
+		sub.m.Lock()
+		// If subscriber is not active any more, remove it.
+		if !sub.active {
+			c.subscribers.remove(sub)
+			sub.m.Unlock()
+			continue
+		}
+
+		if len(sub.c) > 0 {
+			close(sub.c)
+			sub.c = make(chan *WorkloadUpdate, 1)
```

Comment:
Seems like there might be a race here? Workload API handler might read the channel after we check its length... and if there is another read between the time we close the channel and make a new one, we'll get a zero-value update... sounds unlikely though, curious to hear your thoughts

Reply:
Hmm I see. What if I call Updates() function, get the channel, let the buffer fill on it then read from it? If we replace the channel on the subscriber, and close the old one, I have to call Updates() again?

In workload handler, we do <-Updates()? This might be OK for that kind of consumption, but cause problems for others (for instance if I call Updates() and then store the received channel). Maybe a comment on the Updates method would be good, warning that folks should not hold onto the channel they receive from it because it will be replaced?

Reply:
That's right, Updates() returns the current channel. If it is empty, it will block until it receives some element from notifySubscribers() (which is the only place where the channel is written). Once the channel is read it becomes empty again and the next call to Updates() will block until a new update arrives. But what if the channel wasn't read before a new notifySubscribers() happened? Then notifySubscribers() will lock the mutex, close the channel (this won't forbid the reader to get the buffered update) and create a new one where the update will be put.

Yes, I updated workload handler to do <-Updates(). You are absolutely right, I will put a comment on that method.